### PR TITLE
Added functionality to skip Azure certificates

### LINKED ISSUE #872: Optimize certificate reporting in Health Checker
Azure joined servers generate a lot of unnecessary errors as all the "MS-Organization-P2P-Access" are listed as invalid. I'd like to be able to exclude these, or only report on certificates that are assigned to Exchange roles.

These certificates cause a similar issue in the Exchange GIU as all the Azure certs are constantly expiring and triggering unnecessary warnings.

### PR DESCRIPTION
**Pull request by lusassl-msft**
**Description:**
There are several automatic-generated certificates that exist on Azure hosted Exchange servers. These certificates are not bound to any Exchange services but are returned via `Get-ExchangeCertificate`. As some of them are very short-lived (like the `MS-Organization-P2P-Access [YYYY]` certificate with a lifetime of one day), these certificates are generating false-positives when running the Exchange HealthChecker script.

Resolve #872 

**Fix:**
Added a logic to exclude some of the certificates from being returned via `Get-ExchangeServerCertificate` function call. We validate if the `Subject` or `Issuer` matches a regex pattern and if they do, we validate if they're self-signed or not as a second indicator. The exclusion function should be used only in a very limited way and the cases should be documented in the code.

**Validation:**
Lab/Pester

### CODE REVIEW

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1 [hunk 1]
```diff
@@ -3,10 +3,84 @@
...
+            # We found a matching certificate based on subject or issuer, check for IsSelfSigned as second indicator
+            $skiplistEntry = $ExclusionObject[$certificateIndex]
+            if ($skiplistEntry -ne -1) {
+                return ($skiplistEntry.IsSelfSigned -eq $Certificate.IsSelfSigned)
+            } else {
+                return $false
+            }
+        } else {
+            return $false
+        }
```
Comment (dpaulson45) [inline Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1:64]:
```suggestion
        
        $certificateMatch = $Exclusions | Where-Object { $Certficate.Subject -match $_.IorSPattern -or $Certficate.Issuer -match $_.IorSPattern } | Select-Object -First 1
        
        if ($null -ne $certificateMatch) {
           return $certificateMatch.IsSelfSigned -eq $Certificate.IsSelfSigned
        }
        return $false 
```

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1 [hunk 2]
```diff
@@ -3,10 +3,84 @@
...
+                # Exclude "MS-Organization-P2P-Access [YYYY]" certificate with one day lifetime on Azure hosted machines.
+                # See: What are the MS-Organization-P2P-Access certificates present on our Windows 10/11 devices?
+                # https://docs.microsoft.com/azure/active-directory/devices/faq
+                NewCertificateExclusionEntry -IssuerOrSubjectPattern "CN=MS-Organization-P2P-Access \[[12][0-9]{3}\]$" -IsSelfSigned $true
+                # Exclude "DC=Windows Azure CRP Certificate Generator" (TenantEncryptionCertificate)
+                # The certificates are built by the Azure fabric controller and passed to the Azure VM Agent.
+                # If you stop and start the VM every day, the fabric controller might create a new certificate.
+                # These certificates can be deleted. The Azure VM Agent re-creates certificates if needed.
+                # https://docs.microsoft.com/azure/virtual-machines/extensions/features-windows
+                NewCertificateExclusionEntry -IssuerOrSubjectPattern "DC=Windows Azure CRP Certificate Generator" -IsSelfSigned $true
```
Comment (dpaulson45) [inline Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1:81]: Let's clean up the comments to be outside the list.
```suggestion
            # Add the certificates that should be excluded from the Exchange certificate check (we don't return an object for them)
                # Exclude "MS-Organization-P2P-Access [YYYY]" certificate with one day lifetime on Azure hosted machines.
                # See: What are the MS-Organization-P2P-Access certificates present on our Windows 10/11 devices?
                # https://docs.microsoft.com/azure/active-directory/devices/faq
                 # Exclude "DC=Windows Azure CRP Certificate Generator" (TenantEncryptionCertificate)
                # The certificates are built by the Azure fabric controller and passed to the Azure VM Agent.
                # If you stop and start the VM every day, the fabric controller might create a new certificate.
                # These certificates can be deleted. The Azure VM Agent re-creates certificates if needed.
                # https://docs.microsoft.com/azure/virtual-machines/extensions/features-windows
            $certificatesToExclude = @(
                NewCertificateExclusionEntry  "CN=MS-Organization-P2P-Access \[[12][0-9]{3}\]$" $true
                NewCertificateExclusionEntry "DC=Windows Azure CRP Certificate Generator" $true
```
You should be able to get by with two parameters without needing to pass the parameter name.

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1 [hunk 3]
```diff
@@ -3,10 +3,84 @@
 
 . $PSScriptRoot\..\..\..\..\Shared\ErrorMonitorFunctions.ps1
 function Get-ExchangeServerCertificates {
+    param(
+        [Parameter(Mandatory = $false)]
+        [switch]
+        $ExcludeCertificatesOnSkiplist
```
Comment (dpaulson45) [inline Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1:9]: Let's remove the parameter as it isn't needed. 

We also have `$Script:Server` within the script so moving to a share type location would require the removal of that anyways.

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1 [hunk 4]
```diff
@@ -111,24 +190,29 @@ function Get-ExchangeServerCertificates {
...
-                    $certInformationObj | Add-Member -MemberType NoteProperty -Name "IsSanCertificate" -Value $sanCertificateInfo
-                    $certInformationObj | Add-Member -MemberType NoteProperty -Name "Namespaces" -Value $certDnsNameList
-                    $certInformationObj | Add-Member -MemberType NoteProperty -Name "Services" -Value $cert.Services
-                    $certInformationObj | Add-Member -MemberType NoteProperty -Name "IsCurrentAuthConfigCertificate" -Value $isAuthConfigInfo
-                    $certInformationObj | Add-Member -MemberType NoteProperty -Name "LifetimeInDays" -Value $certificateLifetime
-                    $certInformationObj | Add-Member -MemberType NoteProperty -Name "Status" -Value $certStatus
-                    $certInformationObj | Add-Member -MemberType NoteProperty -Name "CertificateObject" -Value $cert
-
-                    $certObject += $certInformationObj
+                    if ($excludeCertificate -eq $false) {
```
Comment (dpaulson45) [inline Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1:193]: Remove this, not sure why we are trying other things in the function if we aren't adding it.

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1 [hunk 5]
```diff
@@ -24,6 +98,11 @@ function Get-ExchangeServerCertificates {
                 try {
                     $certificateLifetime = ([System.Convert]::ToDateTime($cert.NotAfter, [System.Globalization.DateTimeFormatInfo]::InvariantInfo) - (Get-Date)).Days
                     $sanCertificateInfo = $false
+                    $excludeCertificate = $false
+
+                    if ($ExcludeCertificatesOnSkiplist) {
+                        $excludeCertificate = ShouldCertificateBeSkipped -Exclusions $certificatesToExclude -Certificate $cert
+                    }
```
Comment (dpaulson45) [inline Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeServerCertificates.ps1:105]: Basing this off of `$ExcludeCertificatesOnSkiplist` being removed. 

```suggestion
                        $excludeCertificate = ShouldCertificateBeSkipped -Exclusions $certificatesToExclude -Certificate
                        
                        if ($excludeCertficate) {
                        Write-Verbose "Excluding cert ENTER CERT SUBJECT NAME HERE. Moving to next cert"
                        continue
                        }
```

**Review by dpaulson45 — CHANGES_REQUESTED** [review]

**Review by dpaulson45 — DISMISSED** [review]

**Comment by JeremyTBradshaw** [comment]
@lusassl-msft I'm seeing the cert show up still, using current/latest version of HealthChecker.ps1.  The subject is "CN=MS-Organization-P2P-Access [2022]".  Just wondering if I'm missing something or if this didn't turn out successful after the commit?

**Pull request merged**